components/uniset/src/bin/gen_properties.rs

```diff
@@ -0,0 +1,458 @@
+    defaults: &HashMap<String, String>,
+) -> HashMap<String, String> {
+    // create the map of applicable property values, and initialize it to
+    // defaults for the whole Unicode range
+    let mut prop_vals: HashMap<String, String> = HashMap::new();
+    prop_vals.clone_from(defaults);
+
+    // determine if this code point matches any of the block-wide defaults
+    // ("overrides")
+    for (range, block_prop_vals) in blocks {
+        if range.contains(char::from_u32(code_point).unwrap()) {
+            let block_prop_vals_clone = block_prop_vals.iter().map(|(k, v)| (k.clone(), v.clone()));
+            prop_vals.extend(block_prop_vals_clone);
+        }
+    }
+
+    // finally, apply the overrides for this code point
+    for (range, code_point_prop_vals) in code_point_overrides {
+        if range.contains(char::from_u32(code_point).unwrap()) {
+            let code_point_override_prop_vals_clone = code_point_prop_vals
+                .iter()
+                .map(|(k, v)| (k.clone(), v.clone()));
+            prop_vals.extend(code_point_override_prop_vals_clone);
+        }
+    }
+
+    prop_vals
+}
+
+fn get_binary_prop_unisets(
+    prop_aliases: &HashMap<String, HashSet<String>>,
+    code_points: &HashMap<u32, HashMap<String, String>>,
+) -> HashMap<String, UnicodeSet> {
+    let mut m: HashMap<String, UnicodeSet> = HashMap::new();
+
+    for (canonical_name, all_names) in prop_aliases {
+        let mut uniset_builder = UnicodeSetBuilder::new();
+        for (code_point, code_point_prop_vals) in code_points {
+            let code_point_prop_names: HashSet<String> =
+                code_point_prop_vals.keys().cloned().collect();
+            if !all_names.is_disjoint(&code_point_prop_names) {
+                uniset_builder.add_char(char::from_u32(*code_point).unwrap());
+            }
+        }
+        if !&uniset_builder.is_empty() {
+            let uniset = uniset_builder.build();
+            m.insert(canonical_name.clone(), uniset);
+        }
+    }
+
+    m
+}
+
+fn main() {
+    let args: Vec<String> = env::args().collect();
+    let filename = &args[1];
+
+    if let Ok(lines) = read_lines(filename) {
+        let line_opts = lines.map(get_line_str_opt);
+        let line_strs = line_opts.flatten();
+        let parseable_lines = line_strs.filter(|line| !is_skip_ppucd_line(line));
+
+        let mut prop_aliases: HashMap<String, HashSet<String>> = HashMap::new();
+
+        let mut defaults: HashMap<String, String> = HashMap::new();
+
+        // TODO: need advice - how should we not add Hash and Eq from UnicodeSet unless we need it?
+        let mut blocks: HashMap<UnicodeSet, HashMap<String, String>> = HashMap::new();
+
+        let mut code_point_overrides: HashMap<UnicodeSet, HashMap<String, String>> = HashMap::new();
+
+        let mut code_points: HashMap<u32, HashMap<String, String>> = HashMap::new();
+
+        for line in parseable_lines {
+            if is_property_line(&line) {
```

Comment:
Looks like a good use case for an enum match.

https://doc.rust-lang.org/book/ch06-02-match.html#patterns-that-bind-to-values